packages/react-graphql/src/hooks/README.md

```diff
@@ -0,0 +1,174 @@
+
+import ApolloClient from 'apollo-client';
+import {ApolloProvider} from 'react-apollo';
+import {ApolloProvider as ApolloHooksProvider} from 'react-graphql';
+
+const client = new ApolloClient();
+
+function App() {
+  return (
+    <ApolloProvider client={client}>
+      <ApolloHooksProvider client={client}>
+        <MyRootComponent />
+      </ApolloHooksProvider>
+    </ApolloProvider>
+  );
+}
+
+render(<App />, document.getElementById('root'));
+```
+
+## `useQuery`
+
+Here is an example of how to use `useQuery` with a query document.
+
+```tsx
+import React from 'react';
+import {useQuery} from '@shopify/react-graphql';
+
+import customerListQuery from './graphql/CustomerListQuery.graphql';
+
+function CustomerList() {
+  const {data, loading, refetch} = useQuery(customerListQuery);
+  if (loading) {
+    return <div>Loading...</div>;
+  }
+
+  const customers = data && data.customers ? data.customers : [];
+  return (
+    <ul>
+      {customers.map(customer => (
+        <li key={customer.id}>{customer.displayName}</li>
+      ))}
+    </ul>
+    <button onClick={async () => {
+        await refetch();
+    }}>Refetch Customers Data</button>
+  );
+}
+```
+
+You can also use `useQuery` with an async query component created from `@shopify/react-graphql`
+
+Note that because the `.graphql` document is being asynchronously loaded, `refetch` and other helper methods may not be available until the query document is loaded.
+
+```tsx
+import React from 'react';
+import {createAsyncQueryComponent, useQuery} from '@shopify/react-graphql';
+
+const CustomerListQuery = createAsyncQueryComponent({
+  load: () => import('./graphql/CustomerListQuery.graphql'),
+});
+
+function CustomerList() {
+  const {data, loading, refetch} = useQuery(customerListQuery);
+  if (loading) {
+    return <div>Loading...</div>;
+  }
+
+  const customers = data && data.customers ? data.customers : [];
+  return (
+    <ul>
+      {customers.map(customer => (
+        <li key={customer.id}>{customer.displayName}</li>
+      ))}
+    </ul>
+    <button onClick={async () => {
+        if (!refetch) {
+            return;
+        }
+        await refetch();
+    }}>Refetch Customers Data</button>
+  );
+}
+```
+
+## `useMutation`
+
+Returns a function to trigger the mutation, and the result of the mutation run that has the following API.
+
+```ts
+export interface MutationResult<Data = Record<string, any>> {
+  data?: Data;
+  error?: ApolloError;
+  loading: boolean;
+  called: boolean;
+  client: ApolloClient<Object>;
+}
+```
+
+```tsx
+import React from 'react';
+import {Form, TextField, Button, Banner} from '@shopify/polaris';
+import {useQuery} from '@shopify/react-graphql';
+
+import createCustomerMutation from './graphql/CreateCustomerMutation.graphql';
+
+function CustomerDetail() {
+  const [name, setName] = React.useState('');
+
+  let resultBanner= null;
+  const [createCustomer, {loading}] = useMutation(createCustomerMutation, {
```

Comment:
Could we use a promise instead if these `onCompleted` and `onError` callbacks?
```js
const createCustomer = useMutation(createCustomerMutation, options);

try {
  const data = await createCustomer({ variables: { ... } });
} catch{
}
```

Reply:
so there are two separate strategy here

1. give you back the promise to handle yourself => the old HoC way
2. allow the use of `onCompleted` & `onError` so I can take care of the promise for you => the `<Mutation />` way

IMO, I like 2, because the amount of boiler plate for the try catch around mutation.

I am however, curious in what scenario you would prefer to handle this yourself instead of just using the callback?

Reply:
My preference will be `give you back the promise to handle yourself`.

It makes the hook/component more flexible/robust by delegating error handling to the consumer.

Reply:
ok, spoke with @lemonmade to offer just the mutation to try yourself.
I still think its nice to have a way of handling all mutation error the same way, but something that's better done as additional hooks in web only.

Reply:
@michelleyschen - Maybe I'm biased because I've been using apollo-react-hooks for the past couple weeks and that's how mutations are done with this package. But I feel it's a better API overall.

I hate callbacks in general. It hides the execution flow compared to a "simpler" promise/async-await.

I agree that abstracting the error handling is better to keep consistency. How hard would it be to have something similar to the `useQuery`, with a response like `[createCustomer, {loading, errors, data}]`?

The main reason why I don't like having a `onCompleted` explicitly is because you'll have to manage errors in it too, since the `onError` will be specific to http/apollo errors if I'm not mistaken. User errors will have to live inside `onCompleted`, which means you'll have quite a bit of logic there compared to your example. 

I also think that the callback option has a bad API. Based on the example again, you'll have rendering which happen to be in a hook, so consumers will have rendering in several places in the component instead of having it grouped at the end of it (when you return the rendered component). And this will get even worse if the user create a custom hook which uses `useMutation`